### DISCUSSION TITLE
Blazor Example

### DISCUSSION THREAD

**Discussion by T-Alkathiri** [JonPSmith/AuthPermissions.AspNetCore#31] [Ideas]
Thank you very much for this excellent library.
Would you please show an example for integrating Blazor Server/Wasm with this library it will be highly appreciated.
Thanks for your cooperate.

**Comment by JonPSmith** [comment]
Hi @T-Alkathiri,

I'm mainly asked to work on backend code (mainly because of the book I wrote, [Entity Framework Core in Action](https://bit.ly/EfCoreBookEd2)), so I haven't learnt Blazor. If anyone wants to build an example on AuthP with Blazor Server, then that would be nice.

PS. Blazor Wasm isn't really a possibility as the AuthP needs to store data about users, roles etc. that is shared across all the users.

**Reply by T-Alkathiri** [reply]
Hi Jon,
Thank you very much for your reply, but you really scared me with this : PS. Blazor Wasm isn't really a possibility as the AuthP needs to store data about users, roles etc. that is shared across all the users.
Dos that mean I can not use Blazor Wasm as my frontend project, because I am trying to implement your excellent AuthP (Multi Tenant) library web API as a backend and Blzor Wasm as a front end .

**Reply by JonPSmith** [reply]
From what I have read there are two main ways to use Blazor:

1. [Blazor Server](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-6.0#blazor-server) which has a ASP.NET Core application acting as the backend. That works with AuthP.
2. [Blazor WebAssembly (WASM)](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-6.0#blazor-webassembly) where the whole application lives in the browser. That can't use AuthP because there is no way to access the common database.

Does that make sense?

**Reply by T-Alkathiri** [reply]
You are right about Blazor WebAssembly
But my idea is to create a WebApi project acting as backend and using AuthP, and create Blazor wasm acting as frontend (UI's) and consume the WebApi via Https and requesting al the data from Api's end point.
Is this possible or not?
Thank you very much.

**Reply by JonPSmith** [reply]
No, the Roles/Permissions as the `HasPermission` attribute / method won't work which means you can't use the AuthP authentication feature.

**Reply by T-Alkathiri** [reply]
So Bad, That means I can use only Blazor Server if I want to utilize the AuthP authentication feature.

**Reply by JonPSmith** [reply]
If you **really** want to use a API + WASM then you would need to build an authentication method in your WASM to use the enum Permissions. The claims will contain the permissions, so you would need to:

- Make the enum Permissions available for both your WebApi app and the the WASM
- You will need to create an authentication method in your WASM. If you can use ASP.NET Core Policy-based authorization in the WASM, then you need to copy over the code in the [AuthPermissions.AspNetCore.PolicyCode](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/AuthPermissions.AspNetCore/PolicyCode) and register the policy code with ASP.NET Core. If you can't use the Policy-based, then you need to build your own code.
- You would need to copy the code in the [AuthPermissions.BaseCode.PermissionsCode](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/AuthPermissions.AspNetCore/PolicyCode) folder into the WASM.

That's a lot of extra work, but if you want WebApi + WASM then you can do it.

**Reply by trevoralanbrooks** [reply]
@JonPSmith Thank you so much for this amazing library with examples!  I would love to see a Blazor Server example, I'm having a very difficult time translating from its current MVC Views cshtml format to Blazor and Razor Pages, any help would be appreciated!  :-)  (I have the Entity Framework stuff migrated with no issues, but I have to admit, I'm rather sketchy at working with front end technologies.)

**Reply by JonPSmith** [reply]
Hi @trevoralanbrooks,

The AuthP library provides the backend code/database only, but to make / test the backend code I provide various in this repo. Lots of people want a demo of using AuthP with Blazor (see issue #38), but its a LOT of work, so we haven't got a demo.

The consensus in issue #38 is you build a Web API backend and a Blazor WASM frontend. Also, be aware that the Blazor WASM frontend isn't secure, so you are relying on the login/authentication of the Web API backend to control user access. 

PS. The MVC Views are only there to show what you can do - you should look at AuthP's services and features and work out what you need for your project. There is [lots of documentation](https://github.com/JonPSmith/EfCore.TestSupport/wiki) and lots of long article (articles are found to the top right in the docs).

**Reply by trevoralanbrooks** [reply]
Hi @JonPSmith,

I'm definitely working through it and making some painfully slow progress, just basically trying to reproduce your Example3 App as a Blazor Server App.  I appreciate your response and will continue to push through it.  Also, thank you for the links, the docs really help a ton!  I think I've read almost every word and watched every video so far!

Thanks again,
Trevor

**Comment by T-Alkathiri** [comment]
Thank you very much for your clarification.
I think I have to think twice for this because it is really a lot of work I have to do and it could be a lot of errors.
Any how thank you so much for your effort and time.
